nav2_graceful_controller/src/graceful_controller.cpp

```diff
@@ -199,24 +155,109 @@ geometry_msgs::msg::TwistStamped GracefulController::computeVelocityCommands(
       logger_, "Could not transform %s to %s: %s",
       costmap_ros_->getBaseFrameID().c_str(), costmap_ros_->getGlobalFrameID().c_str(),
       ex.what());
-    return cmd_vel;
+    throw nav2_core::NoValidControl("Could not transform");
   }
 
-  // Generate and publish local plan for debugging / visualization
-  nav_msgs::msg::Path local_plan;
-  if (!simulateTrajectory(pose, motion_target, costmap_transform, local_plan, reversing)) {
-    throw nav2_core::NoValidControl("Collision detected in the trajectory");
+  // Compute distance to goal as the path's integrated distance to account for path curvatures
+  double dist_to_goal = nav2_util::geometry_utils::calculate_path_length(transformed_plan);
+
+  // If we've reached the XY goal tolerance, just rotate
+  if (dist_to_goal < goal_dist_tolerance_ || goal_reached_) {
```

Comment:
Why isn't this considering if we want to rotate to the goal using the `prefer_final_rotation` parameter?

Reply:
So, there are really two different features at work here:

1) `prefer_final_rotation` is actually the feature down on L212 - when we are close to the goal (less than max lookahead) and prefer final rotation, we will force the target_pose heading to be the "heading to target pose" (imagine robot is at 0,0,0 and the target_pose is 0.1,0,1.57 - if you feed this into the control law, you take a massive sweeping arc to try and approach that 1.57 heading straight on. Instead, we use atan2(0.0, 0.1) and will end up doing no sweeping turn and then just turning in place at the end). I've found this to be probably the most important feature for actually successfully following the majority of paths regardless of the planner than generated them (and especially true for planners that are not kinematics aware).

2) When you've reached the XY goal, if `prefer_final_rotation` is true you probably have a huge heading error - but even if you aren't preferring final rotation you might have a small heading error (usually, we satisfy the heading requirement before we satisfy XY - but, if your localization is a bit unstable, that may not be the case). In either case, the control law gets quite unstable when the XY error is so small, and it can generate huge swooping motions to correct the heading, so you really ALWAYS want to do that final bit of rotation when you are already sitting on top of your goal.

In some ways, this second part is like the `latch_xy_goal_tolerance` behavior that other controllers have - in my controller, I actually had that parameter, and `goal_reached_` (as this code calls it) wouldn't be set to true, but rather to `latch_xy_goal_tolerance`. Regardless, in the existing graceful_controller outside nav2, you always entered this rotation behavior even if latch and prefer were set to false - you just wouldn't have it be sticky unless latch was true.

Reply:
So, if you don't ever allow an in-place rotation at the end and miss your goal tolerances, you're going to make massive sweeping arcs (well off the original plan) to get to the goal. I can't imagine that is what users of an ackermann robot are going to want.

I can add a parameter to disable ever doing final rotations, but I think my documentation would probably be "if you find yourself wanting this, use a different controller"

Reply:
> I can add a parameter to disable ever doing final rotations, but I think my documentation would probably be "if you find yourself wanting this, use a different controller"

Perhaps that is the right answer, but let me bring this up and if you still think that's the play, I'm OK with that.

If you have an ackermann vehicle, then I imagine alot of those users are going to be using the Smac Planners that generate circular minimum-turning-radius curves to kinematically make sure they get to the goal orientation within their constraints. Do you think that the arc this would generate would be that much more extreme than the arc that the minimum turning radius circle would create that the carrot on the path plan would be following?

You have a better intuition on this planner, so I default to your thoughts

Reply:
Two thoughts on this:

 * I'm not sure there is a way to actually enforce a "minimum turning radius" with this control law. I can't think of a straight forward way to derive k1/k2 based on a minimum turning radius...
 * While this controller may not generate more "extreme" arcs, they certainly aren't kinematically aware (since the derivation of the control law is based on differential drive). I would think if you've gone through the effort of configuring and tuning a kinematically aware planner, you want to rely on that for planning larger motions (rather than letting a "dumber" controller just kinda make stuff up).

It just seems like this is the wrong tool for an Ackerman robot. I've seen it used on diff drive, holonomic, semi-holonomic powered casters (it really shines here because you mostly act like a diff drive and don't make crazy movements), and a walking robot - but not something with harder constraints like Ackerman.

Reply:
OK - works for me. It just wouldn't surprise me if a ticket came in in a few years wanting to use it or a PR making it parameterized, but we can address it if/when that time comes. 

I think if you make the note in the readme / config guide / [navigation algorithms table](https://docs.nav2.org/plugins/index.html#controllers) about which robot types it supports, we're good to go. 

> I'm not sure there is a way to actually enforce a "minimum turning radius" with this control law. I can't think of a straight forward way to derive k1/k2 based on a minimum turning radius...

Just as a note: Pure Pursuit used by the DARPA grand challenge and alot of Nav2 ackermann users is also not kinematically aware. It works 'good enough' even without promises

Reply:
Mm, the control law is derived from the unicycle model because it was intended to be used in a differentially driven wheelchair. So I don't think it was intended to be used in other kinds of vehicles, especially ackermans.

With this in mind, the control law explodes when your dist_to_goal -> 0, so skipping the target heading or just rotating when reaching the goal is necessary. This is the big problem with this controller.

> I think if you make the note in the readme / config guide / [navigation algorithms table](https://docs.nav2.org/plugins/index.html#controllers) about which robot types it supports, we're good to go.

When I wrote the docs for the controller, I only specified it for differential drive: https://docs.nav2.org/plugins/index.html#controllers